File: src/random.c
#include <math.h>
#include <stdint.h>
#include <stdlib.h>
#include <time.h>

/* A C-program for TT800 : July 8th 1996 Version */
/* by M. Matsumoto, email: matumoto@math.keio.ac.jp */
/* genrand() generate one pseudorandom number with double precision */
/* which is uniformly distributed on [0,1]-interval */
/* for each call.  One may choose any initial 25 seeds */
/* except all zeros. */

/* See: ACM Transactions on Modelling and Computer Simulation, */
/* Vol. 4, No. 3, 1994, pages 254-266. */

#define N 25
#define M 7

uint64_t x[N]; /* initial 25 seeds, change as you wish */

static double genrand() {
  uint64_t y;
  static int k = 0;
  static uint64_t mag01[2] = {
      0x0, 0x8ebfd028 /* this is magic vector `a', don't change */
  };
  if (k == N) { /* generate N words at one time */
    int kk;
    for (kk = 0; kk < N - M; kk++) {
      x[kk] = x[kk + M] ^ (x[kk] >> 1) ^ mag01[x[kk] % 2];
    }
    for (; kk < N; kk++) {
      x[kk] = x[kk + (M - N)] ^ (x[kk] >> 1) ^ mag01[x[kk] % 2];
    }
    k = 0;
  }
  y = x[k];
  y ^= (y << 7) & 0x2b5b2500;  /* s and b, magic vectors */
  y ^= (y << 15) & 0xdb8b0000; /* t and c, magic vectors */
  y &= 0xffffffff;             /* you may delete this line if word size = 32 */
                               /*
                                  the following line was added by Makoto Matsumoto in the 1996 version
                                  to improve lower bit's correlation.
                                  Delete this line to o use the code published in 1994.
                               */
  y ^= (y >> 16);              /* added to the 1994 version */
  k++;
  return ((double)y / (unsigned long)0xffffffff);
}

/**
 * Initialize the random seeds.
 */
void Randomize() {
  srand(time(NULL));
  for (int i = 0; i < N; i++) {
    x[i] = rand();
  }
}

/**
 * Gives a random float evenly distributed in the interval [min, max]
 */
float RanFl(float min, float max) {
  return genrand() * (max - min) + min;
}

/**
 * Gives a random int evenly distributed in the interval [min, max-1]
 */
int RanInt(int min, int max) {
  double ran;
  do
    ran = genrand();
  while (ran == 1);
  return floor(ran * (max - min) + min);
}

/**
 * Returns true with a probability of prob
 * (i.e. if prob is is 1 always return true, if prob is 0 always return false)
 */
int RanProb(float prob) {
  return genrand() <= prob;
}

File: src/random.zig
//! random.zig: a simple psuedorandom number generator
//! The original source used this random number generator written by
//! M. Matsumoto. See the original source for more details. This is a
//! version translated to Zig.
//!
//! It isn't clear if this specific algorithm is needed, so for now we
//! will use it and later see if it can be replaced with a Zig PRNG.

const std = @import("std");
const math = std.math;

const Prng = std.rand.DefaultPrng;

/// Constants for random number generation
const N: u32 = 25;
const M: u32 = 7;
const magic = [2]u64{ 0x0, 0x8ebf_d028 };

/// Initial random seeds for the random number generator
var seeds: [N]u64 = undefined;

/// Generate the next random number in the sequence using the algorithm from M. Matsumoto.
var k: i32 = 0;
fn nextRandom() f64 {
    if (k == N) {
        var kk: u32 = 0;
        while (kk < N - M) : (kk += 1) {
            seeds[kk] = seeds[kk + M] ^ (seeds[kk] >> 1) ^ magic[seeds[kk] % 2];
        }
        while (kk < N) : (kk += 1) {
            seeds[kk] = seeds[kk - 18] ^ (seeds[kk] >> 1) ^ magic[seeds[kk] % 2];
        }
        k = 0;
    }

    var y = seeds[@intCast(usize, k)];
    y ^= (y << 7) & 0x2b5b_2500; // s and b, magic vectors
    y ^= (y << 15) & 0xdb8b_0000; // t and c, magic vectors
    y &= 0xffffffff;

    y ^= (y >> 16);
    k += 1;

    return @intToFloat(f64, y) / @intToFloat(f64, 0xffff_ffff);
}

/// Initialize the random number seeds
pub fn init(initial_seed: u64) void {
    var prng = Prng.init(initial_seed);
    var rand = prng.random();

    for (seeds) |*seed| {
        seed.* = rand.int(u64);
    }
}

/// Return a random float evenly distributed in the interval [min, max]
pub fn randomFloat(min: f32, max: f32) f32 {
    return nextRandom() * (max - min) + min;
}

/// Return a random int evenly distributed in the interval [min, max-1]
pub fn randomInt(min: i32, max: i32) i32 {
    const min_f = @intToFloat(f64, min);
    const max_f = @intToFloat(f64, max);

    var r = nextRandom();
    while (r == 1) r = nextRandom();
    return @floatToInt(i32, r * (max_f - min_f) + min_f);
}

/// Returns true with a probability of p
/// (i.e. if p is is 1 always return true, if p is 0 always return false)
pub fn randomProb(p: f32) bool {
    return nextRandom() <= p;
}

const expect = std.testing.expect;

test "random" {
    // init with a deterministic seed
    init(2);

    // expect these values
    try expect(randomInt(0, 100) == 80);
    try expect(randomInt(0, 100) == 45);
    try expect(randomInt(0, 100) == 86);
    try expect(randomInt(0, 100) == 79);
    try expect(randomInt(0, 100) == 41);
    try expect(randomInt(0, 100) == 84);
    try expect(randomInt(0, 100) == 65);
    try expect(randomInt(0, 100) == 29);
    try expect(randomInt(0, 100) == 49);
    try expect(randomInt(0, 100) == 62);
}
